feat: GeometryId check for Acts::Experimental::Detector

ssdetlab opened the pull request:
Adding GeometryIdentier checks to Acts::Experimental::Detector. Surface Ids are being checked not only for the uniqueness, but also for being assigned in the first place, allowing to avoid undefined GeoId cases. 

Similar checks are added for the DetectorVolume GeoIds. These are, for example, used in the MaterialMapping algo, and the duplicates, as well as undefined Ids, may cause the mapping to malfunction. Using a placeholder unordered_map to check for duplicates.  

In all cases throwing invalid_argument, as it is done in already implemented duplicate check for Surfaces.

Planning to add similar checks for Portals.

asalzburger commented:
Excellent to see this appear, I will have a look.

asalzburger commented:
@ssdetlab can you please apply the clang-format 

In order to rest assured that you use the right version, you can simply run the docker version, i.e. if you have a docker client running, apply

```sh
./CI/check_format_local .
```

in your cloned repository.

asalzburger commented:
And: wow, most of the unit tests show that this is really not well done:

```sh
The following tests FAILED:
	 11 - CylindricalDetectorHelper (Failed)
	 12 - Detector (Failed)
	 13 - DetectorBuilder (Failed)
	 18 - IndexedRootVolumeFinderBuilder (Failed)
	112 - DetectorVolumeFinders (Subprocess aborted)
	115 - DetectorNavigator (Failed)
	277 - DetectorJsonConverter (Failed)
```

ssdetlab commented:
Added geometry id assignment to the  CylindricalVolumeHelper test. Also testing the clang-format. Let's see, if it's going to work.

asalzburger commented on Tests/UnitTests/Core/Detector/CylindricalDetectorHelperTests.cpp:
```diff
@@ -534,6 +565,16 @@ BOOST_AUTO_TEST_CASE(ProtoContainerZR) {
+    GeometryIdGenerator generator(
+        generatorConfig, Acts::getDefaultLogger("SequentialIdGenerator",
+                                                Acts::Logging::VERBOSE));
+    auto cache = generator.generateCache();
```
Geometry generators can have a cache, e.g. if you want to assign a container type generator to several volume builders, you could can increase the layer counter in the cache, e.g.

ssdetlab commented:
Went through failing Unit Tests and added geometry Id assignment to DetectorVolumes where needed. The only non-trivial change is in the Json DetectorVolume writer. There I added the lines that write to/read from Json the Volume's geometry Id.

asalzburger commented:
Hi @ssdetlab - please apply the clang-format again (as written above).

asalzburger commented:
We are almost there, 1 UnitTest is still failing, let's fix this one & then move forward.

asalzburger reviewed (changes requested):
This is excellent work, with one more Unittest fixed, I can approve this in the next round.

ssdetlab commented:
Added the fixing lines to the Json writer. Will try to figure out what's wrong with the formatting

asalzburger commented:
> Added the fixing lines to the Json writer. Will try to figure out what's wrong with the formatting

Format look good now!

acts-project-service commented:
## :white_check_mark: Athena integration test results [392612e2a45d9f96985fdcfb1e9e97b2eb105427]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368275)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368274)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368273)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368272)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368271)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368270)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368269)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368268)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368267)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368266)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368265)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368264)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368263)||
|:green_circle: | [run_ci_tests: ../athena/Tracking/Acts/ActsConfig/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368262)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368261)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368260)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368259)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368258)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368257)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368256)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368255)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368254)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35368253)||

asalzburger commented:
Congratulations @ssdetlab - your first PR went in successfully! Great work.
